tests/components/broadlink/__init__.py

```diff
@@ -196,12 +196,13 @@ def patch_discovery(mock_discovery):
     """Patch device discovery."""
     broadcast_addrs = list(mock_discovery)
     devices = list(mock_discovery.values())
+    return_values = [iter(dev_list) for dev_list in devices]
 
     with patch(
         "homeassistant.components.broadlink.get_broadcast_addrs",
         return_value=broadcast_addrs,
     ), patch(
-        "homeassistant.components.broadlink.discovery.blk.discover",
-        side_effect=devices,
+        "homeassistant.components.broadlink.discovery.blk.xdiscover",
+        side_effect=return_values,
```

Comment:
I'm stuck here. I don't know why the tests are not passing, maybe I should look elsewhere? I will do it, but if anyone can help me find a solution things will move faster.

Reply:
The tests are not passing because of:

### test_discovery.py
test_setup_discover_update_ip_address
test_setup_discover_update_hostname
test_setup_discover_do_not_change_hostname

When we call `await hass.data[DOMAIN].discovery.coordinator.async_refresh()` we go into some kind of dead lock. Something is blocking the execution. No idea what. I'll have to dig deeper. This is consuming me time :(

Reply:
```
______ ERROR at teardown of test_setup_discover_update_ip_address[pyloop] ______
[gw0] linux -- Python 3.7.1 /__w/core/core/venv/bin/python

loop = <_UnixSelectorEventLoop running=False closed=True debug=False>
hass_storage = {'core.config_entries': {'data': {'entries': [{'connection_class': 'unknown', 'data': {'host': '192.168.1.128', 'mac':...g_room_remote', 'last_changed': '2020-11-19T19:24:01.680166+00:00', ...}}], 'key': 'core.restore_state', 'version': 1}}
request = <SubRequest 'hass' for <Function test_setup_discover_update_ip_address[pyloop]>>

    @pytest.fixture
    def hass(loop, hass_storage, request):
        """Fixture to provide a test instance of Home Assistant."""
    
        def exc_handle(loop, context):
            """Handle exceptions by rethrowing them, which will fail the test."""
            # Most of these contexts will contain an exception, but not all.
            # The docs note the key as "optional"
            # See https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.call_exception_handler
            if "exception" in context:
                exceptions.append(context["exception"])
            else:
                exceptions.append(
                    Exception(
                        "Received exception handler without exception, but with message: %s"
                        % context["message"]
                    )
                )
            orig_exception_handler(loop, context)
    
        exceptions = []
        hass = loop.run_until_complete(async_test_home_assistant(loop))
        orig_exception_handler = loop.get_exception_handler()
        loop.set_exception_handler(exc_handle)
    
        yield hass
    
        loop.run_until_complete(hass.async_stop(force=True))
        for ex in exceptions:
            if (
                request.module.__name__,
                request.function.__name__,
            ) in IGNORE_UNCAUGHT_EXCEPTIONS:
                continue
            if isinstance(ex, ServiceNotFound):
                continue
>           raise ex

tests/conftest.py:142: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/local/lib/python3.7/asyncio/events.py:88: in _run
    self._context.run(self._callback, *self._args)
/usr/local/lib/python3.7/asyncio/futures.py:344: in _set_state
    _copy_future_state(other, future)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

source = <Future at 0x7fda06d054a8 state=finished raised StopIteration>
dest = <Future pending cb=[_chain_future.<locals>._call_check_cancel() at /usr/local/lib/python3.7/asyncio/futures.py:348, <1 more>, _wait.<locals>._on_completion() at /usr/local/lib/python3.7/asyncio/tasks.py:440]>

    def _copy_future_state(source, dest):
        """Internal helper to copy state from another Future.
    
        The other Future may be a concurrent.futures.Future.
        """
        assert source.done()
        if dest.cancelled():
            return
        assert not dest.done()
        if source.cancelled():
            dest.cancel()
        else:
            exception = source.exception()
            if exception is not None:
>               dest.set_exception(exception)
E               TypeError: StopIteration interacts badly with generators and cannot be raised into a Future
```

Reply:
I wrapped the generator with a `try: except` clause and realized that there is a StopIteration being propagated outside the generator in the spoiled tests.
```python3
    def discover(self, broadcast_addrs, current_entries):
        """Discover Broadlink devices on the given networks, ignoring known devices."""
        known_devices = {
            (get_ip_or_none(entry.data[CONF_HOST]), entry.data[CONF_MAC])
            for entry in current_entries
        }
        try:
            for addr in broadcast_addrs:
                for device in blk.xdiscover(
                    discover_ip_address=addr, timeout=self.timeout
                ):
                    host, mac_addr = device.host[0], device.mac.hex()
                    if (host, mac_addr) in known_devices:
                        continue

                    if device.type not in SUPPORTED_TYPES:
                        continue

                    self.create_flow(device)
        except StopIteration:
            _LOGGER.error("STOP ITERATION")
```
```
DEBUG:homeassistant.components.broadlink.updater:Finished fetching Living Room (RM mini 3 at 192.168.1.128) data in 0.001 seconds
ERROR:homeassistant.components.broadlink.discovery:STOP ITERATION
```
This fixes the deadlock. I think the answer is [PEP479](https://www.python.org/dev/peps/pep-0479/).

Reply:
I managed to fix it. Goodbye bug. I won.